### ISSUE TITLE
MSAL double redirects - loading state

### ISSUE THREAD

**Issue by szopqa** [AzureAD/microsoft-authentication-library-for-js#6244]
### Core Library

MSAL.js (@azure/msal-browser)

### Core Library Version

2.38.0

### Wrapper Library

MSAL Angular (@azure/msal-angular)

### Wrapper Library Version

2.5.9

### Public or Confidential Client?

Public

### Description

Hi, as per the discussion [here](https://github.com/AzureAD/microsoft-authentication-library-for-js/issues/2514), I'm trying to implement the redirection flow in my SPA using B2C. Still, I would also like to implement the loading state on my main page after the redirection happens, before navigating the user to the protected route. Here's our setup:

https://example.com/ -> marketing site with the `Log in` button (using B2C user flow)
https://example.com/foo -> SPA where the users should land after they complete the sign-up/in process

We also use [HashLocationStrategy](https://angular.io/api/common/HashLocationStrategy), so registering `https://example.com/foo` (which in fact is `https://example.com/#/foo`) as a redirect URL is not possible, as OAuth does not support fragmented URLs.

The main question I have is how to handle the "loading" state on the homepage after MSAL redirects to it before I can manually redirect the user to `/#/foo` after successful login. 
Here's what I tried:

1. I tried the approach when the app would know that it's being used in the redirect flow by `state` being present in the URL. I modified the routing module, so it redirects from `#state...` to separate `LoginRedirectComponent` which by default only renders the loader

```
const routes: Routes = [
    { path: '', redirectTo: 'home', pathMatch: 'full' },
    { path: 'state', redirectTo: 'login-redirect', pathMatch: 'full' },
    { path: 'error', redirectTo: 'home' },
    { path: 'home', component: HomeComponent },
    {
        path: 'foo',
        component: FooComponent,
        canActivate: [MsalGuard],
    },
    {
        path: 'login-redirect',
        component: LoginRedirectComponent,
        pathMatch: 'full',
        canActivate: [MyGuard],
    },
]

@NgModule({
    imports: [
        RouterModule.forRoot(routes, {
            useHash: true,
            initialNavigation:
                !BrowserUtils.isInIframe() && !BrowserUtils.isInPopup()
                    ? 'enabledBlocking'
                    : 'disabled',
        })
    ],
    exports: [RouterModule],
})
export class AppRoutingModule {}
```

I also implemented custom `MyGuard` which is extracting `window.location.hash` as it didn't happen by default within the router for some reason

```
@Injectable({
    providedIn: 'root',
})
export class MyGuard implements CanActivate {
    constructor(private dataService: DataService) {}

    canActivate(
        route: ActivatedRouteSnapshot,
        state: RouterStateSnapshot
    ): boolean | UrlTree {
        const hash = window.location.hash

        if (hash && hash !== '') {
            this.dataService.setData(hash)
        }
        return true
    }
}
```

2. I implemented `LoginRedirectComponent` to act as a sign in/up loader, handling some additional post-sign-up logic we have, which extracts the mentioned hash and executes `handleRedirectPromise` with it. In order to make it work I had to use extracted hash by the dedicated service.

```
async ngOnInit(): Promise<void> {
        await this.msalService.instance.handleRedirectPromise(
            this.dataService.getData()
        )

        this.msalBroadcastService.inProgress$
            .pipe(
                filter(
                    (status: InteractionStatus) =>
                        status === InteractionStatus.None
                ),
                takeUntil(this._destroying$)
            )
            .subscribe(() => {
                this.checkAndSetActiveAccount()
            })
    }
```

What I am observing RN is that indeed the loader is rendering, it does all the things it should and eventually redirects to `/foo`, but after a short period of time I am anyway redirected to the `/` which is the HomeComponent, which I really would like to avoid. I tried playing around with `navigateToLoginRequestUrl` but it seems to make no difference. 

My questions then are:
 - Why am I seeing the redirect happening to the HomeComponent even when I execute `handleRedirectPromise` on the `LoginRedirectComponent`? Can that be somehow avoided?
 - Is it possible to implement such a flow with hash-based routing, where the usage of a separate `redirectUrl` page is not possible? I was able to implement what I want very quickly without it.
 - Is there any other way I could tell my Home component that it should render the loader as it's being used in the redirection flow

Thanks in advance!

### MSAL Configuration

```javascript
{
    auth: {
        clientId: "clientId",
        authority: "authority",
        knownAuthorities: [authorityDomain],
        redirectUri: "http://localhost:4200",
        navigateToLoginRequestUrl: true
    },
    cache: {
        cacheLocation: BrowserCacheLocation.LocalStorage,
        storeAuthStateInCookie: isIE,
    },
    system: {
        loggerOptions: {
            loggerCallback: (logLevel, message) => {
                logger.logErrorOnApplicationInsight(new Error(message))
            },
            logLevel: environment.msalLoggingLevel,
            piiLoggingEnabled: false,
        },
    },
}
```

### Identity Provider

Azure B2C Basic Policy

### Source

External (Customer)

**Comment by tnorling** [comment]
This is happening because after the login the service will redirect you to https://localhost:4200#code=..., MSAL sees this and removes the code (which triggers the change to your home route).

Try passing in the url you want to end up on (e.g. http://localhost:4200/#your_desired_route) as the redirectUri. Yes, this is not supported by the protocol, but MSAL will strip off the hash, store the value in temporary storage, and set the redirectUri to http://localhost:4200 (so this is what needs to be registered in the portal) and upon return, instead of removing the #code= it should _replace it_ with #your_desired_route.

**Comment by szopqa** [comment]
Hi @tnorling , thanks for your reply.

I only tried registering them in B2C, but after that failed I haven't tried using them as the redirect URL
![Screenshot 2023-07-20 at 10 46 54](https://github.com/AzureAD/microsoft-authentication-library-for-js/assets/16049905/6c0c64a5-99a0-4747-a4dc-3fda7f19bb0c)


I've just updated the usage of `loginRedirect` as you suggested, but I am seeing exactly the same behavior - I am taken to the loader component, and then back to the home component.
![Screenshot 2023-07-20 at 11 25 08](https://github.com/AzureAD/microsoft-authentication-library-for-js/assets/16049905/bf060455-fd90-4723-b150-b7109d0bc077)
After analyzing the logs it looks like the redirect to home (see the empty hash log from home.component.ts) happens when `handleRedirectPromise` is still being processed in the `login-redirect-component`, so I again see the loader for a moment, then the Homepage and eventually the `foo` component.
